caffe2/python/examples/cifar10_fp16.py

```diff
@@ -0,0 +1,210 @@
+import numpy as np
+import os
+import time
+from timeit import default_timer as timer
+
+from caffe2.python import brew, core, cnn, net_drawer, workspace
+from caffe2.proto import caffe2_pb2
+from caffe2.python.model_helper import ModelHelper
+
+def AddInput(model, batch_size, db, db_type, param_type="float", is_test=False):
+    """Adds the data input part."""
+    reader = model.CreateDB(
+        "reader_"+db,
+        db=db,
+        db_type=db_type)
+
+    # Note: Using C2 datum for this test
+    data, label = model.ImageInput(
+        [reader], ["data", "label"],
+        use_gpu_transform=True,
+        batch_size=batch_size, use_caffe_datum=True,
+        mean=128., scale=32, crop=32, mirror=1, color=3,
+        is_test=is_test, output_type=param_type
+    )
+    data = model.StopGradient(data, data)
+
+    return data, label
+
+# Move conv1 to fp16 only
+def AddCifar10Model(model, data, param_type):
+
+    with brew.arg_scope([brew.conv, brew.fc], dtype="float16"):
+        conv1 = brew.conv(model, data, 'conv1', 3, 32, 5,
+                weight_init=('GaussianFill', {'std':0.0001, 'mean' : 0.0}), pad=2)
+        pool1 = brew.max_pool(model, conv1, 'pool1', kernel=3, stride=2) #14 or 15
+        relu1 = brew.relu(model, pool1, "relu1")
+        conv2 = brew.conv(model, relu1, 'conv2', 32, 32, 5, weight_init=('GaussianFill', {'std' : 0.01}), pad=2)
+        conv2 = brew.relu(model, conv2, conv2)
+        pool2 = brew.average_pool(model, conv2, 'pool2', kernel=3, stride=2) #5 or 6
+        conv3 = brew.conv(model, pool2, 'conv3', 32, 64, 5, weight_init=('GaussianFill', {'std' : 0.01}), pad=2)
+        conv3 = brew.relu(model, conv3, conv3)
+        pool3 = brew.average_pool(model, conv3, 'pool3', kernel=3, stride=2)
+        fc1 = brew.fc(model, pool3, 'fc1', 64 * 3 * 3, 64, weight_init=('GaussianFill', {'std' : 0.1}))
+        fc2 = brew.fc(model, fc1, 'fc2', 64, 10, weight_init=('GaussianFill', {'std' : 0.1}))
+
+    # Final cast out -> fp32
+    if param_type == "float16":
+        fc2 = model.HalfToFloat(fc2, "fc2_fp32")
+
+    softmax = model.Softmax(fc2, 'softmax')
+
+    return softmax
+
+def AddAccuracy(model, softmax, label):
+    """Adds an accuracy op to the model"""
+
+    accuracy = model.Accuracy([softmax, label], "accuracy", top_k=1)
+    return accuracy
+
+def AddTrainingOperators(model, softmax, label, using_snapshot, use_fp16=False):
+    """Adds training operators to the model."""
+    xent = model.LabelCrossEntropy([softmax, label], 'xent')
+    loss = model.AveragedLoss(xent, "loss")
+
+    AddAccuracy(model, softmax, label)
+
+    print("adding gradient ops")
```

Comment:
Please use logging with debugging verbosity level (logger.info) instead of prints.